docs/proposals/26-workload-identity.md

```diff
@@ -0,0 +1,514 @@
+the only component having write access to this CRD, while extension controllers
+and other components relying on workload identities should request no more than
+read access. This `WorkloadIdentityBinding` CRD will be reconciled by a
+dedicated controller named `workloadidentity-refresher` from Gardenlet
+controller manager responsible to request OIDC tokens and write them into the
+target store. The workload identity `spec.config` will be also written into the
+target store.
+
+At the time this GEP is written, it is envisioned only secrets to be used as
+store targets. The token will be available on the data key `wiToken` in the
+secret, while the service provider config will be at `wiConfig`. These secrets
+will be labeled with
+`workloadidentity.security.gardener.cloud/provider=(aws|gcp|...)` so that the
+extensions can easily select them and make adjustments via admission webhooks,
+e.g. transform the service provider config and the token into canonical form
+usable by the respective service provider SDK.
+
+The secret `cloudprovider` will be reused as target store when the shoot is
+using workload identity tokens as infrastructure credentials, the flow should be
+as follow:
+
+1. Shoot reconciler creates/updates a `WorkloadIdentityBinding` resource in the
+   seed based on the configuration of the `WorkloadIdentity` and the shoot spec.
+   The name of the `WorkloadIdentityBinding` resource will be `cloudidentity`
+   while the name of the target store secret `cloudprovider`.
+1. `workloadidentity-refresher` receives event for `WorkloadIdentityBinding`
+   resource. It parses the token from the target store and if the token is still
+   valid, it just applies the provider config to the target store, otherwise a
+   new token is requested from the Gardener API and written to the store
+   together with the provider config.
+1. The status of the `WorkloadIdentityBinding` is updated with the "issued at"
+   and "expires at" timestamps of the token.
+1. The `WorkloadIdentityBinding` is requeued for reconciliation again when the
+   token will be suitable for renewal.
+
+A sample `WorkloadIdentityBinding.extensions.gardener.cloud` resource will look like:
+
+```yaml
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: WorkloadIdentityBinding
+metadata:
+  name: cloudidentity
+  namespace: shoot--local--foo
+spec:
+  type: aws
+  config:
+    apiVersion: aws.security.gardener.cloud/v1alpha1
+    kind: Config
+    iamRoleARN: arn:aws:iam::112233445566:role/gardener-dev
+  contextObject: # Optional field, various metadata about context of use of the token
+    apiVersion: core.gardener.cloud/v1beta1
+    kind: Shoot
+    name: foo
+    namespace: garden-local
+    uid: 54d09554-6a68-4f46-a23a-e3592385d820
+  workloadIdentity:
+    name: banana-testing
+    namespace: garden-local
+  store:
+    secretRef:
+      name: cloudprovider
+status:
+  issuedAt: 2024-02-09T08:35:02Z
+  expiresAt: 2024-02-09T16:35:02Z
+```
```

Comment:
Even after reading this, I still struggle to understand why we need this CRD. I would prefer to follow a similar approach as with the token requestor controller, see https://github.com/gardener/gardener/blob/master/example/resource-manager/30-secret-tokenrequestor.yaml for example.
- The "target store" seem to always be a secret (what else could it be?), and the `type/{provider}Config/contextObject/workloadIdentity` fields are "clear" since such a secret is created in a shoot namespace (i.e., can be derived from the shoot spec).
- The token-requestor write `serviceaccount.resources.gardener.cloud/token-renew-timestamp: "<...>"` annotation which is a good fit to replace the `status` in your proposal (if you insist, you can also write `{issued,expires}-at`, though, it's probably only of limited use compared to additional space it consumes.

Reply:
> The "target store" seem to always be a secret (what else could it be?),

A CSI driver, for example.

> and the type/{provider}Config/contextObject/workloadIdentity fields are "clear" since such a secret is created in a shoot namespace (i.e., can be derived from the shoot spec).

Please, have in mind that while the main purpose of the GEP is to provide alternative to the static infrastructure credentials of a shoot cluster, it is designed to be usable in other use cases, like DNS, backup, certificate, etc. controllers. 

Do you imply that these controllers must be watching the shoot or seed resource from the Gardener API now? 

Secret name is also not exposed as configuration in the shoot spec because shoot owners can easily misconfigure it and cause more harm than good. Also, no naming convention is proposed, because any change in the convention must be coordinated between all controllers which is tedious work. Changing the store type also would need to be coordinated. Therefore the store type and name are written in the resource and any controller can read it from there.

> I would prefer to follow a similar approach as with the token requestor controller, see https://github.com/gardener/gardener/blob/master/example/resource-manager/30-secret-tokenrequestor.yaml for example.


This is very subjective, during the recent years Gardener started relying too much on annotations basically for everything. And annotations are proven to not follow any API best practices, it is extremely hard to learn and understood the hundreds of annotations in what resources are applicable and what are their effect. This GEP certainly will not take the annotations path.

Reply:
> A CSI driver, for example.

I don't think I have the full picture to follow this thought, but as the doc already mentions, it seems that `Secret`s will be the only "target store" for the foreseeable future. Do you have concrete plans w.r.t. CSI drivers, or is this just a theoretical idea that likely will never be implemented? Otherwise, please extend the doc with more details about this approach.

> This is very subjective, during the recent years Gardener started relying too much on annotations basically for everything.

Is this statement any less subjective?

> And annotations are proven to not follow any API best practices, it is extremely hard to learn and understood the hundreds of annotations in what resources are applicable and what are their effect.

Annotating and labelling existing resources is a best-practice in the Kubernetes community. Dozens of projects, including "first-class citizens" like cloud-controller-managers, but also well-adopted SIG projects like [external-dns](https://github.com/kubernetes-sigs/external-dns) make use of this approach. Reading a document to understand which labels/annotations have which effect isn't much different from reading a document to understand which spec fields have which effect.

My point is that a lot of this token handling for service providers is very similar to what we have already implemented w.r.t. token handling for API servers/`ServiceAccount`s. It makes a lot of sense to align the two approaches/implementations and not invent another approach just because it was implemented by other people.

To make it concrete for the examples in this doc, the `gardenlet` could create the following `Secret`:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: cloudprovider
  namespace: shoot--local--foo
  annotations:
    workloadidentity.authentication.gardener.cloud/name: <value of `.workloadIdentityRef.name` in the `SecretBinding` referenced `.spec.secretBinding` in the `Shoot`>
    workloadidentity.authentication.gardener.cloud/namespace: <value of `.workloadIdentityRef.namespace` in the `SecretBinding` referenced `.spec.secretBinding` in the `Shoot`>
  labels:
    authentication.gardener.cloud/purpose: workloadidentity-token-requestor
type: Opaque
```

Now, the controller in `gardenlet` responsible for requesting tokens can fetch the `Shoot` this namespace belongs to, create a `security.gardener.cloud/v1alpha1.TokenRequest`, and sync the token into the `.data.token`. There is no need to replicate the workload identity spec into the seed clusters.

For extensions, `gardenlet` could create similar `Secret`s named with the type of the extension:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: workloadidentity-some-extension
  namespace: shoot--local--foo
  annotations:
    workloadidentity.authentication.gardener.cloud/name: <value of `.spec.extensions[].workloadIdentity.name` for the extension of type `some-extension` in the `Shoot`
    workloadidentity.authentication.gardener.cloud/namespace: <namespace of the `Shoot`>
  labels:
    authentication.gardener.cloud/purpose: workloadidentity-token-requestor
type: Opaque
```

This makes the token handling much simpler and aligns it with the existing `token-requestor` controller, the indirection via the CRD is not needed.

> In my drafts initially I was thinking to use secret with annotations and labels, but it looks like every controller in the seed have read/write access to all secrets in the seed, hence every controller would be able to request workload identity tokens, which we want to avoid.

This "problem" exists already today with the `ServiceAccount` token secrets. All these controllers that have access to `Secret`s today will keep their access also after this GEP has been implemented. In your proposal, the token is eventually synced into a `Secret`, so again, all controllers will be able to fetch it. In this light, the fact that they cannot request new tokens doesn't help much since token generation is anyways limited to the workload identity referenced in the `Shoot` (hence, the token will be always valid for the same service provider, no matter whether newly requested or read from a "target store" secret).
If you still think is a problem, it should probably be addressed separately such that it also applies for the `ServiceAccount` token secrets (could perhaps be done via a validating webhook for `Secret`s in `gardener-resource-manager` that checks that such `Secret`s are only written by gardenlet or extensions).

Aligning the approach with the existing `token-requestor` also has the benefit that it might be reusable in the garden cluster when managed by `gardener-operator`, similar to how we have made the `ServiceAccount` token request approach reusable for the virtual garden cluster.
We are currently discussing how we can automate management of DNS records and the backup bucket for the garden cluster (currently, these entities are not covered by `gardener-operator` and must be managed with separate tooling). Details are not clear yet, but a CRD in the `extensions.gardener.cloud` API group probably invalidates the reusability option.

Reply:
I think we should give @rfranzke's proposal a shot: it's streamlined and easy, no CRD detour. Do you see any real issues with it after seeing the example object?

Reply:
Addressed with 53ef0d243, please take a look.


> Aligning the approach with the existing `token-requestor` also has the benefit that it might be reusable in the garden cluster when managed by `gardener-operator`, similar to how we have made the `ServiceAccount` token request approach reusable for the virtual garden cluster.

I am not sure I can follow here, but it looks more natural the Kubernetes service account tokens of the runtime cluster to be used. It might not even work, because Workload Identity will be available only after the Gardener API server is up and running, but the DNS and the backup must be setup at that point of time, looks like some cyclic dependency would be introduced.